### ISSUE TITLE
C code cannot be compiled with --threads:on

### ISSUE THREAD

**Issue by bluenote10** [nim-lang/Nim#2620]
When using `--threads:on` I cannot compile even a simple `echo "Hello World"`. The C compilation fails with:

```
.../nimcache/stdlib_system.c: In function ‘atomicinc_14401’:
.../nimcache/stdlib_system.c:6097:41: error: ‘__ATOMIC_RELAXED’ undeclared (first use in this function)
.../nimcache/stdlib_system.c:6097:41: note: each undeclared identifier is reported only once for each function it appears in
.../nimcache/stdlib_system.c: In function ‘atomicdec_14406’:
.../nimcache/stdlib_system.c:6104:41: error: ‘__ATOMIC_RELAXED’ undeclared (first use in this function)
```

Version: latest devel 26eae7d00e73c65670775091bad8bfd796b3e1f1

This was the true cause of #2619.

**Comment by bluenote10** [comment]
It turns out that this problem is caused by the GCC version, which is 4.6.3 in my case. The atomic builtins referenced in `atomic.nim` are only available with GCC 4.7+. I was trying to implement `AtomMemModel` as `enum` (as it was in [this earlier version](https://github.com/Araq/Nim/blob/30823c1ce3992d48251069af48ed9d26b1238ba4/lib/system/atomics.nim)), but this still leaves undeclared identifiers for all the function calls.

**Comment by dom96** [comment]
Can this be closed?

**Comment by bluenote10** [comment]
From my point of view: Yes, problem solved by using to a more recent GCC.

But it probably depends on which GCC versions to target. Not sure what the plans are in general. I remember Araq consider fixing it. GCC 4.6 is still the default for Ubuntu 12.04, which may still be around until 2017 (where support ends), so we can expect others to run into this. I guess the simplest solution is to deprecate GCC 4.6, mention it in the manual, and produce a meaningful error "GCC version too old". The real issue here is probably just the error message, since it is rather difficult to see the actual cause behind these "undeclared" errors.

**Comment by dom96** [comment]
I'll leave this open so that we can hopefully improve the error message.

**Comment by Varriount** [comment]
How would we do that? The error occurs when handling control of compilation over to gcc.

**Comment by dom96** [comment]
@Varriount many different options. Off the top of my head: generate a macro in the C code which checks for the presence of ``__ATOMIC_RELAXED`` (and creates a nice error when it doesn't exist), get the Nim compiler to run ``gcc -v`` to check the version of gcc on the system.

**Comment by bluenote10** [comment]
Closing this in favor of #5992
